### PR TITLE
[Logger] Use lifecycle callbacks for logging

### PR DESCRIPTION
**Pull request by rahul-tuli**
- Add preliminary changes that were missed
- Propagate step while logging loss
- Fix WANDB logging
- Use LifeCycleCallbacks for logging loss and model info
- Update tests

Test Script:
```python
from sparseml.transformers import train

model = "Xenova/llama2.c-stories15M"
teacher_model = "Xenova/llama2.c-stories15M"
dataset_name = "open_platypus"
concatenate_data = False
output_dir = "./output_finetune"
recipe = "/home/ubuntu/projects/sparseml/local/recipes/llama_gmp.yaml"
num_train_epochs=2
overwrite_output_dir = True
splits = {
    "train": "train[:10%]",
}

train(
    model=model,
    distill_teacher=teacher_model,
    dataset=dataset_name,
    output_dir=output_dir,
    recipe=recipe,
    num_train_epochs=num_train_epochs,
    overwrite_output_dir=overwrite_output_dir,
    concatenate_data = concatenate_data,
    splits = splits
)
```

Recipe: 
```yaml
# llama_gmp.yaml
test_oneshot_stage:
  obcq_modifiers:
    SparseGPTModifier:
      sparsity: 0.7
      block_size: 128
      sequential_update: False
      quantize: False
      percdamp: 0.01
      prunen: 0
      prunem: 0
      targets: [
        "model.layers.0"
      ]
      target_ids: ["attention_mask", "position_ids"]  
test_train_stage:
  pruning_modifiers:
    ConstantPruningModifier:
      targets: [
        "re:.*self_attn.q_proj",
        "re:.*self_attn.k_proj",
        "re:.*self_attn.v_proj",
        "re:.*self_attn.o_proj",
        "re:.*mlp.gate_proj",
        "re:.*mlp.up_proj"
      ]
      start: 0
```

WANDB run: wandb: 🚀 View run laced-sponge-38 at: https://wandb.ai/neuralmagicml/sparseml-local_scripts/runs/5ak0wzdk

Log file has been attached with the description:
[27-02-2024_12.38.02.log](https://github.com/neuralmagic/sparseml/files/14419596/27-02-2024_12.38.02.log)

### CODE REVIEW

File: src/sparseml/core/session.py [hunk 1]
```diff
@@ -325,22 +324,25 @@ def _log_loss(self, event_type: EventType, loss: Any):
...
-        if loss is not None and log_ready(
-            current_log_step=current_step,
-            last_log_step=self._last_loss_log_step,
-            log_frequency=self.state.loggers.log_frequency,
-            check_model_update=check_model_update,
-        ):
-            loss = loss if isinstance(loss, dict) else {"loss": loss}
+        # always log loss if available
+        if loss is not None:
+            loss = loss if isinstance(loss, dict) else {"Loss": loss}
```
Comment (bfineran) [inline src/sparseml/core/session.py:342]:
```suggestion
            loss = loss if isinstance(loss, dict) else {"loss": loss}
```

**Review by bfineran — APPROVED** [review]
